spring-integration-core/src/main/java/org/springframework/integration/handler/SuspendingMessageHandler.java

```diff
@@ -0,0 +1,132 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.integration.handler;
+
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+
+import org.springframework.integration.IntegrationMessageHeaderAccessor;
+import org.springframework.integration.aggregator.CorrelationStrategy;
+import org.springframework.integration.aggregator.HeaderAttributeCorrelationStrategy;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessagingException;
+
+/**
+ * A pair of message handlers; the first suspends the thread until the second receives
+ * a corresponding message.
+ *
+ * @author Gary Russell
+ * @since 4.2
+ *
+ */
+public class SuspendingMessageHandler extends AbstractReplyProducingMessageHandler {
+
+	private final ConcurrentHashMap<Object, BlockingQueue<Message<?>>> suspensions =
+			new ConcurrentHashMap<Object, BlockingQueue<Message<?>>>();
+
+	private final ConcurrentHashMap<Object, Thread> inProcess = new ConcurrentHashMap<Object, Thread>();
+
+	private final long timeout;
+
+	private final CorrelationStrategy correlationStrategy;
+
+	public SuspendingMessageHandler(long timeout) {
+		this(new HeaderAttributeCorrelationStrategy(IntegrationMessageHeaderAccessor.CORRELATION_ID), timeout);
+	}
+
+	public SuspendingMessageHandler(CorrelationStrategy correlationStrategy,
+			long timeout) {
+		this.correlationStrategy = correlationStrategy;
+		this.timeout = timeout;
+	}
+
+	@Override
+	protected Object handleRequestMessage(Message<?> requestMessage) {
+		Object key = SuspendingMessageHandler.this.correlationStrategy.getCorrelationKey(requestMessage);
+		if (key == null) {
+			throw new MessagingException(requestMessage, "Correlation Strategy returned null");
+		}
+		Thread existing = this.inProcess.putIfAbsent(key, Thread.currentThread());
+		if (existing != null) {
+			throw new MessagingException(requestMessage, "Correlation key ("
+					+ key + ") is already in use by " + existing.getName());
+		}
+		BlockingQueue<Message<?>> blockingQueue = createOrObtainQueue(key);
+		try {
+			Message<?> release = blockingQueue.poll(this.timeout, TimeUnit.MILLISECONDS);
+			if (release != null) {
+				this.suspensions.remove(key);
+				if (release.getPayload() instanceof Throwable) {
+					throw new MessagingException(requestMessage, "Releasing flow returned a throwable",
+							(Throwable) release.getPayload());
+				}
+				else {
+					return buildResult(requestMessage, release);
+				}
+			}
+			else {
+				synchronized(blockingQueue) {
+					if (!blockingQueue.offer(requestMessage)) {
+						// second chance
+						return buildResult(requestMessage, blockingQueue.poll());
+					}
+				}
+				return null;
```

Comment:
I think just return `null` isn't enough for this component.
How about `MessageTimeoutException`?
From other side we can regulate the final behavior with some options. E.g. `requiresReply`.
WDYT?